File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.cc
#include<iostream>
#include<algorithm>
using namespace std;

int board[501][501];
int dp[501][501];
int dx[4] = {0,-1,0,1};
int dy[4] = {1,0,-1,0};
int size;

int dfs(int row, int col){
    if(dp[row][col]>0){
        return dp[row][col];
    }

    int cnt = 1;
    for(int i=0; i<4; i++){
        int nRow = row + dx[i];
        int nCol = col + dy[i];
        if(nRow < 0 || nRow>=size||nCol<0 || nCol>=size) continue;
        if(board[nRow][nCol]> board[row][col]){
            cnt = max(cnt, 1 + dfs(nRow, nCol));
        }
    }

    return dp[row][col] = cnt;
}

int main(void){
    ios::sync_with_stdio(0);
    cin.tie(0);
    cin >> size;
    for(int i=0; i<size; i++){
        for(int j=0; j<size; j++){
            cin >> board[i][j];
        }
    }


    int ans = 0;
    for(int i=0; i<size; i++){
        for(int j=0; j<size; j++){
            ans = max(ans, dfs(i,j));
        }
    }

    cout << ans <<"\n";
    return 0;
}

File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.py
import sys
from functools import lru_cache
sys.setrecursionlimit(10**9)
input = sys.stdin.readline
size = int(input())
board = [list(map(int, input().split())) for _ in range(size)]
dp = [[0 for _ in range(size)] for _ in range(size)]
def dfs(row, col):
    if dp[row][col]:
        return dp[row][col]
    
    cnt = 1
    for n_row, n_col in [(row-1,col), (row+1,col), (row,col-1), (row, col+1)]:
        if 0<=n_row<size and 0<=n_col<size and board[row][col] < board[n_row][n_col]:
            cnt = max(cnt, dfs(n_row, n_col) + 1)
    
    dp[row][col] = cnt
    return cnt
        
ans = 0
for i in range(size):
    for j in range(size):
        ans = max(ans, dfs(i,j))
print(ans)